README.md
# RTL
[![Gem Version](https://badge.fury.io/rb/rtl.svg)](https://rubygems.org/gems/rtl)
[![Build Status](https://travis-ci.org/abarrak/rtl.svg?branch=master)](https://travis-ci.org/abarrak/rtl)
[![Test Coverage](https://api.codeclimate.com/v1/badges/e6e976b150ddabbc4665/test_coverage)](https://codeclimate.com/github/abarrak/rtl/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/e6e976b150ddabbc4665/maintainability)](https://codeclimate.com/github/abarrak/rtl/maintainability)

This gem helps you check whether a language direction is right-to-left (RTL) or left-to-right (LTR) implicitly, by supplying its full name, iso code, or other options.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'rtl'
```

And then execute:

```sh
$ bundle
```

Or install it yourself as:

```sh
$ gem install rtl
```

## Usage

#### Query a language by string or symbol:

```ruby
require 'rtl'

Rtl.rtl? 'ar'                      #=> true
Rtl.rtl? :ara                      #=> true
```

#### Query a language by ISO 639 code, (the default) ..

```ruby
Rtl.rtl? :syc, :iso_code           #=> true
Rtl.rtl? :eng                      #=> false
```

#### Or by ISO 15924 code .. 

```ruby
Rtl.rtl? 'Hung', :iso_long_code    #=> true
```

#### Or by ISO number ..

```ruby
Rtl.rtl? 130, :iso_number          #=> true
```

#### Or by unicode alias ..
```ruby
Rtl.rtl? 'Hebrew', :unicode_alias  #=> true
```

#### Or by full name ..

```ruby
Rtl.rtl? :Persian, :full_name      #=> true
```

#### Or check all ways ..

```ruby
Rtl.rtl? :kur, :all      #=> true
```

Which is the default, in case no scheme criteria is specified.


#### Get List of all languages by any of the above criteria:

```ruby
Rtl.rtl_languages
#=> ["ar", "ara", "arc", "ae", "ave", "egy", "he", "heb", "nqo", "pal", "phn", "sam", "syc", "syr", "fa", "per", "fas", "ku", "kur"]

Rtl.rtl_languages :iso_long_code
#=> ...
```

## Useful Links
* [ISO 639-2 Language Code List - Codes for the representation of names of languages (Library of Congress).](https://www.loc.gov/standards/iso639-2/php/code_list.php)
* [List of ISO 639-1 codes on Wikipedia.](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)
* [Right-to-left on Wikipedia.](https://en.wikipedia.org/wiki/Right-to-left)
* [ISO 15924 on Wikipedia.](https://en.wikipedia.org/wiki/ISO_15924)
* [Questions & Answers: Which languages are written right-to-left (RTL)?](http://www.i18nguy.com/temp/rtl.html)

## Contributing

Bug reports and pull requests are very much appreciated at [Github](https://github.com/abarrak/rtl).


## License
The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



CHANGELOG
# RTL Gem Changelog

## [0.6.0] - 2022-06-14

- Add support for Yiddish language [@richardjaldin](https://github.com/richardjaldin).

## [0.5.0] - 2022-04-14

- Add support for Pashto (Pushto) [@asped](https://github.com/asped).

## [0.4.0] - 2021-07-24

- Make querying by :all criteria as the default for convenience instead of iso_code only [@abarrak](https://github.com/abarrak).
- Bug fix on error reporting in case incorrect scheme option is passed [@abarrak](https://github.com/abarrak).

## [0.3.0] - 2021-07-24

- Update development dependencies to latest versions [@abarrak](https://github.com/abarrak).

## [0.2.4] - 2019-03-1

- Add support for Urdu [@benwbrum](https://github.com/benwbrum).

## [v0.2.3] - 2016-10-3

- Initial release [@abarrak](https://github.com/abarrak).



test/rtl_test.rb
require 'test_helper'

class RtlTest < Minitest::Test
  include Rtl

  def setup
    @iso_codes        = Rtl::ISO_CODES
    @iso_long_codes   = Rtl::ISO_LONG_CODES
    @iso_numbers      = Rtl::ISO_NUMBERS
    @unicode_aliases  = Rtl::UNICODE_ALIASES
    @full_names       = Rtl::FULL_NAMES
    @wrong_lang       = ["en", "fr", "de", "Spanish", 420]
  end

  def test_respond_corretly_for_iso_codes
    @iso_codes.each do |code|
      assert Rtl.rtl? code, :iso_code
      assert Rtl.rtl? code.to_sym, :iso_code
    end

    @wrong_lang.each { |w| assert !Rtl.rtl?(w) }
  end

  def test_respond_corretly_for_iso_long_codes
    @iso_long_codes.each do |long_code|
      assert Rtl.rtl? long_code, :iso_long_code
      assert Rtl.rtl? long_code.to_sym, :iso_long_code
    end

    @wrong_lang.each { |w| assert !Rtl.rtl?(w, :iso_long_code) }
  end

  def test_respond_corretly_for_iso_numbers
    @iso_numbers.each do |iso_num|
      assert Rtl.rtl? iso_num, :iso_number
      assert Rtl.rtl? iso_num.to_sym, :iso_number
    end

    @wrong_lang.each { |w| assert !Rtl.rtl?(w, :iso_number) }
  end

  def test_respond_corretly_for_unicode_aliases
    @unicode_aliases.each do |u_alias|
      assert Rtl.rtl? u_alias, :unicode_alias
      assert Rtl.rtl? u_alias.to_sym, :unicode_alias
    end

    @wrong_lang.each { |w| assert !Rtl.rtl?(w, :unicode_alias) }
  end

  def test_respond_corretly_for_full_names
    @full_names.each do |full_name|
      assert Rtl.rtl? full_name, :full_name
      assert Rtl.rtl? full_name.to_sym, :full_name
    end

    @wrong_lang.each { |w| assert !Rtl.rtl?(w, :full_name) }
  end

  def test_can_make_exhasutive_search
    [@iso_codes, @iso_long_codes, @iso_numbers, @unicode_aliases, @full_names].each do |set|
      set.each do |el|
        assert Rtl.rtl? el, :all
        assert Rtl.rtl? el.to_sym, :all
      end
    end

    @wrong_lang.each { |w| assert !Rtl.rtl?(w, :all) }
  end

  def test_get_iso_codes_languages
    assert_equal Rtl.rtl_languages, @iso_codes
    assert_equal Rtl.rtl_languages(:iso_code), @iso_codes
  end

  def test_get_iso_long_codes_languages
    assert_equal Rtl.rtl_languages(:iso_long_code), @iso_long_codes
  end

  def test_get_iso_numbers_languages
    assert_equal Rtl.rtl_languages(:iso_number), @iso_numbers
  end

  def test_get_unicode_aliases_languages
    assert_equal Rtl.rtl_languages(:unicode_alias), @unicode_aliases
  end

  def test_get_full_names_languages
    assert_equal Rtl.rtl_languages(:full_name), @full_names
  end

  def test_reports_invalid_scheme_is_passed
    [:my_langs, :iso_special].each do |sch|
      assert_raises(ArgumentError) { Rtl.rtl_languages sch }
    end
  end

  def test_raises_error_in_case_incorrect_scheme_is_passed
    [:my_codes, :iso_777].each do |sch|
      assert_raises(ArgumentError) { Rtl.rtl? :en, sch }
    end
  end
end


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

require 'simplecov'
SimpleCov.start

require 'minitest/autorun'
require 'minitest/pride'
require 'rtl'


lib/rtl/core.rb
module Rtl
  # Formal full names for RTL languages.
  FULL_NAMES = ["Adlam", "Arabic", "Arabic (Nastaliq variant)", "Imperial Aramaic", "Avestan",
                "Cypriot", "Egyptian demotic", "Egyptian hieratic", "Hatran", "Hebrew",
                "Old Hungarian (Hungarian Runic)", "Indus (Harappan)", "Kharoshthi", "Lydian",
                "Mandaic, Mandaean", "Manichaean", "Mende Kikakui", "Meroitic Cursive",
                "Meroitic Hieroglyphs", "Old North Arabian (Ancient North Arabian)",
                "Nabataean", "N’Ko", "Old Turkic, Orkhon Runic", "Palmyrene",
                "Inscriptional Pahlavi", "Psalter Pahlavi", "Book Pahlavi", "Phoenician",
                "Inscriptional Parthian", "Samaritan", "Old South Arabian", "Syriac",
                "Syriac (Estrangelo variant)", "Syriac (Western variant)",
                "Syriac (Eastern variant)", "Thaana", "Woleai", "Persian", "Kurdish", "Urdu",
                "Pashto (Pushto)", "Yiddish"]

  # ISO 639 codes for RTL languages.
  # If a language has 639-1 and 639-2 codes, both are included.
  ISO_CODES = ["ar", "ara", "arc", "ae", "ave", "egy", "he", "heb", "nqo", "pal", "phn", "sam",
               "syc", "syr", "fa", "per", "fas", "ku", "kur", "ur", "urd", "pus", "ps", "yi", "yid"]

  # ISO 15924 codes for RTL languages.
  ISO_LONG_CODES = ['Adlm', 'Arab', 'Aran', 'Armi', 'Avst', 'Cprt', 'Egyd', 'Egyh', 'Hatr',
                    'Hebr', 'Hung', 'Inds', 'Khar', 'Lydi', 'Mand', 'Mani', 'Mend', 'Merc',
                    'Mero', 'Narb', 'Nbat', 'Nkoo', 'Orkh', 'Palm', 'Phli', 'Phlp', 'Phlv',
                    'Phnx', 'Prti', 'Samr', 'Sarb', 'Syrc', 'Syre', 'Syrj', 'Syrn', 'Thaa',
                    'Wole']

  # ISO 15924 numbers for RTL languages.
  ISO_NUMBERS = ['166', '160', '161', '124', '134', '403', '070', '060', '127', '125', '176',
                 '610', '305', '116', '140', '139', '438', '101', '100', '106', '159', '165',
                 '175', '126', '131', '132', '133', '115', '130', '123', '105', '135', '138',
                 '137', '136', '170', '480']

  # Unicode alias names for RTL languages.
  # Not all ISO languages has this alias.
  UNICODE_ALIASES = ["Adlam", "Arabic", "Imperial Aramaic", "Avestan", "Cypriot",
                     "Hatran", "Hebrew", "Old Hungarian", "Kharoshthi", "Lydian", "Mandaic",
                     "Manichaean", "Mende Kikakui", "Meroitic Cursive", "Meroitic Hieroglyphs",
                     "Old North Arabian", "Nabataean", "NKo", "Old Turkic", "Palmyrene",
                     "Inscriptional Pahlavi", "Psalter Pahlavi", "Phoenician",
                     "Inscriptional Parthian", "Samaritan", "Old South Arabian", "Syriac",
                     "Thaana"]

  # Query whether a language is rtl or not.
  def self.rtl? language, scheme = :all
    sch = scheme.to_sym
    l = language.to_s.strip

    case sch
      when :iso_code
        self.iso_codes l
      when :iso_long_code
        self.iso_long_codes l
      when :iso_number
        self.iso_numbers l
      when :unicode_alias
        self.unicode_aliases l
      when :full_name
        self.full_names l
      when :all
        self.all l
      else
        raise ArgumentError.new "Unknown scheme value #{scheme}."
    end
  end

  # Get a list of languages by scheme
  def self.rtl_languages scheme = :iso_code
    sch = scheme.to_s
    s = sch.end_with?('s') ? 'ES' : 'S'
    member = "#{sch.upcase}#{s}".to_sym

    begin
      self.const_get member
    rescue NameError
      raise ArgumentError.new "Unknown Supplied scheme #{scheme}."
    end
  end

  private

    def self.iso_codes lang
      ISO_CODES.include? lang.downcase
    end

    def self.iso_long_codes lang
      ISO_LONG_CODES.include? lang.capitalize
    end

    def self.iso_numbers lang
      ISO_NUMBERS.include? lang
    end

    def self.full_names lang
      FULL_NAMES.map(&:capitalize).include? lang.capitalize
    end

    def self.unicode_aliases lang
      UNICODE_ALIASES.map(&:downcase).include? lang.downcase
    end

    def self.all lang
      iso_codes(lang)  || iso_long_codes(lang) || iso_numbers(lang) ||
      full_names(lang) || unicode_aliases(lang)
    end
end
